packages/vitest/src/node/reporters/dot.ts

```diff
@@ -1,53 +1,182 @@
-import type { UserConsoleLog } from '../../types/general'
+import type { Custom, File, TaskResultPack, TaskState, Test } from '@vitest/runner'
+import type { Vitest } from '../core'
+import { getTests } from '@vitest/runner/utils'
+import c from 'tinyrainbow'
 import { BaseReporter } from './base'
-import { createDotRenderer } from './renderers/dotRenderer'
+import { WindowRenderer } from './renderers/windowedRenderer'
+import { TaskParser } from './task-parser'
+
+interface Icon {
+  char: string
+  color: (char: string) => string
+}
 
 export class DotReporter extends BaseReporter {
-  renderer?: ReturnType<typeof createDotRenderer>
+  private summary?: DotSummary
 
-  onTaskUpdate() {}
+  onInit(ctx: Vitest) {
+    super.onInit(ctx)
 
-  onCollected() {
     if (this.isTTY) {
-      const files = this.ctx.state.getFiles(this.watchFilters)
-      if (!this.renderer) {
-        this.renderer = createDotRenderer(files, {
-          logger: this.ctx.logger,
-        }).start()
-      }
-      else {
-        this.renderer.update(files)
-      }
+      this.summary = new DotSummary()
+      this.summary.onInit(ctx)
+    }
+  }
+
+  onTaskUpdate(packs: TaskResultPack[]) {
+    this.summary?.onTaskUpdate(packs)
+
+    if (!this.isTTY) {
+      super.onTaskUpdate(packs)
     }
   }
 
-  async onFinished(
-    files = this.ctx.state.getFiles(),
-    errors = this.ctx.state.getUnhandledErrors(),
-  ) {
-    await this.stopListRender()
-    this.ctx.logger.log()
+  onWatcherRerun(files: string[], trigger?: string) {
+    this.summary?.onWatcherRerun()
+    super.onWatcherRerun(files, trigger)
+  }
+
+  onFinished(files?: File[], errors?: unknown[]) {
+    this.summary?.onFinished()
     super.onFinished(files, errors)
   }
+}
+
+class DotSummary extends TaskParser {
+  private renderer!: WindowRenderer
+  private tests = new Map<Test['id'], TaskState>()
+  private finishedTests = new Set<Test['id']>()
+
+  onInit(ctx: Vitest): void {
+    this.ctx = ctx
+
+    this.renderer = new WindowRenderer({
+      logger: ctx.logger,
+      getWindow: () => this.createSummary(),
+    })
+
+    this.ctx.onClose(() => this.renderer.stop())
+  }
 
-  async onWatcherStart() {
-    await this.stopListRender()
-    super.onWatcherStart()
+  onWatcherRerun() {
+    this.tests.clear()
+    this.renderer.start()
   }
 
-  async stopListRender() {
-    this.renderer?.stop()
-    this.renderer = undefined
-    await new Promise(resolve => setTimeout(resolve, 10))
+  onFinished() {
+    const finalLog = formatTests(Array.from(this.tests.values()))
+    this.ctx.logger.log(finalLog)
+
+    this.tests.clear()
+    this.renderer.finish()
   }
 
-  async onWatcherRerun(files: string[], trigger?: string) {
-    await this.stopListRender()
-    super.onWatcherRerun(files, trigger)
+  onTestFilePrepare(file: File): void {
```

Comment:
I just noticed we are still using the old API here... We need to experiment with the new one to see what's missing. Otherwise, it will never be out of the experimental

I don't know if we should do it here, but keep in mind that it needs to be rewritten before 2.2 (I can do it myself, if no one has resources)

Reply:
You mean the `File` part, as in we should use `state.getReportedEntity` instead?

When building the summary reporter I had at first that one in use and also some new Reporter API hooks. But as nothing was working well I started to reduce amount of changes. First fix all reporter flickering and scrolling issues and then look into new APIs.